numpy/lib/function_base.py

```diff
@@ -4498,7 +4500,12 @@ def quantile(a,
     if a.dtype.kind == "c":
         raise TypeError("a must be an array of real numbers")
 
-    q = np.asanyarray(q)
+    # Use dtype of array if possible (e.g., if q is a python int or float).
+    if isinstance(q, (int, float)) and a.dtype.kind == "f":
+        q = np.asanyarray(q, dtype=np.result_type(a, q))
```

Comment:
We could also just use this stanza for `percentile` to keep things more uniform.

Note that this would not work for user types either given the check for `a.dtype.kind == "f"`, but I did have to include that since otherwise an example with a datetime broke.